keps/sig-cli/NNNN-kustomize-components.md

```diff
@@ -0,0 +1,308 @@
+- [Proposal](#proposal)
+  - [User Story](#user-story)
+  - [Notes/Constraints/Caveats (optional)](#notesconstraintscaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+(**FIXME:** Left here as a guideline, remove it before sending the PR)
+
+> **ACTION REQUIRED:** In order to merge code into a release, there must be an
+> issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+> milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+> of the targeted release**.
+>
+> For enhancements that make changes to code or processes/procedures in core
+> Kubernetes i.e., [kubernetes/kubernetes], we require the following Release
+> Signoff checklist to be completed.
+>
+> Check these off as they are completed for the Release Team to track. These
+> checklist items _must_ be updated for the enhancement to be released.
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+(**FIXME:** Explain what is the current situation and then roughly explain what this feature does)
+
+> This section is incredibly important for producing high quality user-focused
+> documentation such as release notes or a development roadmap.  It should be
+> possible to collect this information before implementation begins in order to
+> avoid requiring implementors to split their attention between writing release
+> notes and implementing the feature itself.  KEP editors, SIG Docs, and SIG PM
+> should help to ensure that the tone and content of the `Summary` section is
+> useful for a wide audience.
+>
+> A good summary is probably at least a paragraph in length.
+
+## Motivation
```

Comment:
Here is a draft for the **Motivation** section:

Since Kubernetes 1.14 one can use kubectl to quickly spin their
applications on Kubernetes, simply by grouping YAML manifests into base and
variant overlays. Typically, plain kubectl apply -k works well for
applications that support numbered settings and target specific users and
environments.

However, this does not seem to be the case for more complex scenarios that deal
with multivariate applications, i.e., ones that:

support numerous opt-in features and configuration options
target dissimilar audiences with different needs and scopes (developers,
admins, customers, testers etc.)
run on multiple cloud platforms (on-prem, Minikube, GKE, EKS, AKS, etc.)
tailor distinct parts of a very same, parent configuration simulataneously
The community experience shows that, in the real world, declaratively managing
applications that support a lot of optional features and configuration options
with stock Kubernetes CLI tools is quite hard and can quickly get out of
control. The main reason for this is that, by design, kustomize is inclined
towards inheritance rather than composition, which more or less determines how
variants are implemented and used.

Zooming in, the mechanism that kustomize uses to interpret and ultimately
apply patches does not allow for mutating different aspects of the very same
base resource in parallel without modifying its GVKN (unique identifier). That
is, sibling variants that sit in the same level of the hierarchy and are not
chained (i.e., one does not import the other as base) cannot modify their
common parent due to resource ID conflicts, even if they are semantically
distinct (the one might configure app's authentication via LDAP and the other
might configure app's data persistence in an external database).

In addition, following kustomize's inheritance model to represent and
manage applications with a higher degree of freedom in terms of configuration
is rather inefficient. Since these applications essentially rely on mixins to
generate their installation manifests to provide as many off-the-shelf
combinations as possible we end up with duplicate YAML documents in multiple
places. This greatly hinders reusability of variants across deployments and, in
the long term, readability and maintainability. Ideally, we would like to
reduce repetition of variants and treat each variant as a single, unambiguous
element according to the [DRY] principle.

The above raise the challenge of designing a clean and concise solution that,
preserves the core principles of Kubernetes, interferes as little as possible
with the internals of kustomize and, albeit, provides missing flexibility
to users.

Reply:
I have the following comments on this section:

> Since Kubernetes 1.14 one can use kubectl to quickly spin their
> applications on Kubernetes, simply by grouping YAML manifests into base and
> variant overlays. Typically, plain kubectl apply -k works well for
> applications that support numbered settings and target specific users and
> environments.

I suggest we remove this paragraph, since readers of the KEP should already be acquainted with Kustomize. If they're not, we can simply point to the respective KEP.

> However, this does not seem to be the case for more complex scenarios that deal
> with multivariate applications, i.e., ones that:
> ...
> as a single, unambiguous
> element according to the [DRY] principle.

I understand that in these paragraphs, you're basically saying:

* What does Kustomize support now? (inheritance model)
* What are some examples that don't fit to this model? (bullet list)
* How could we make these examples easier to write? (composition model)
* Why can't we do so right now? (explain issue with Kustomize's mechanism)

This is exactly what we need on this section, so we're good. I have only two comments:

1. I think it's easier for the reader to grasp what you're saying if you introduce the composition model earlier on, instead of trying to explain a limitation of Kustomize first, and then showing how the composition model bumps into it.
2. I feel there's some redundant info in these sections that we can trim/simplify, to make them easier to read. Some examples:

    * *"The community experience shows that..."*: Let's simply state that in an inheritance model, you have to copy the same logic across overlays.
    * *"even if they are semantically distinct..."*: You have already provided some examples on why it makes sense to patch a resource multiple times. No need to provide one here as well.

> The above raise the challenge of designing a clean and concise solution that,
> preserves the core principles of Kubernetes, interferes as little as possible
> with the internals of kustomize and, albeit, provides missing flexibility
> to users.

I think that this tailpiece fits more in a blog post about this feature, than a technical document. I understand that you want to wrap the above up somehow, but there's no need. There are quite a few sections that follow below after all.

Reply:
> I suggest we remove this paragraph, since readers of the KEP should already be acquainted with Kustomize. If they're not, we can simply point to the respective KEP
I thought of this too, so I will remove it.

> I think it's easier for the reader to grasp what you're saying if you introduce the composition model earlier on, instead of trying to explain a limitation of Kustomize first, and then showing how the composition model bumps into it.
ACK, I will first refer to the composition concept and then name why this fails to solve certain cases.

> I feel there's some redundant info in these sections that we can trim/simplify, to make them easier to read. Some examples:
I agree for the first bullet so I will rephrase this as needed. I am not sure about the second one though. Do you feel this is covered by the last entry of the bullet list?

> I think that this tailpiece fits more in a blog post about this feature, than a technical document. I understand that you want to wrap the above up somehow, but there's no need. There are quite a few sections that follow below after all.
ACK

Reply:
Here is a second draft for the **Motivation** section based on above comments:

By design, Kustomize and its variant system follow an inheritance model rather
than a composition one. This poses certain limitations in complex deployment
scenarios that deal with multivariate applications, i.e., ones that:

- ship with numerous opt-in features and corresponding configuration options 
- target dissimilar audiences with different needs and scopes (developers,
  admins, customers, testers etc.) 
- run on multiple cloud platforms (on-prem, Minikube, GKE, EKS, AKS, etc.)
- tailor semantically distinct parts of a
  common, parent configuration simulataneously

Currently, declarative management of such applications with stock Kubernetes
CLI tools is quite hard and can quickly get out of control.

Zooming in, the mechanism that Kustomize uses to interpret and ultimately apply
patches does not allow for mutating different aspects of the very same base
resource in parallel without modifying its GVKN (unique identifier). That is,
sibling variants that operate on the same level of the hierarchy and are not
chained (i.e., one does not import the other as base) cannot modify their
common parent due to resource ID conflicts.

Aside, representation and management of applications with a higher degree of
freedom with an inheritance model seems rather inefficient. The reason for this
is that some applications rely on mixins to provide as many off-the-shelf
combinations (sets of manifests) as possible which,foreseeably, leads to
duplicate YAML documents and logic in multiple places. This greatly hinders
reusability of variants across deployments and, in the long term, readability
and maintainability. Ideally, we would like to reduce repetition of variants
and treat each variant as a single, unambiguous element according to the
[DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle.

### Goals

- Introduce the notion of components in ``kustomize`` and, eventually in
  ``kubectl``
- Provide the implementation that allows users to define components, i.e.,
  portable variants that are able to modify a common, base resources without
  conflicts, since patches are serialized
- Allow for efficient and more concise kustomization of multidimensional
  applications by importing reusable variants
- Maintain existing, stable interfaces that end-users are already familiar with,
  i.e., offer components as an optional feature

### Non-Goals

- Overshadow `resources` with `components`
-Abolish inheritance over composition in Kustomize